resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsCache.java

```diff
@@ -289,33 +306,60 @@ void add(DefaultDnsCacheEntry e) {
 
                         // Create a new List for COW semantics
                         List<DefaultDnsCacheEntry> newEntries = new ArrayList<DefaultDnsCacheEntry>(entries.size() + 1);
-                        DefaultDnsCacheEntry replacedEntry = null;
-                        for (int i = 0; i < entries.size(); i++) {
+                        int i = 0;
+                        do {
                             DefaultDnsCacheEntry entry = entries.get(i);
                             // Only add old entry if the address is not the same as the one we try to add as well.
                             // In this case we will skip it and just add the new entry as this may have
                             // more up-to-date data and cancel the old after we were able to update the cache.
                             if (!e.address().equals(entry.address())) {
                                 newEntries.add(entry);
-                            } else {
-                                assert replacedEntry == null;
-                                replacedEntry = entry;
                             }
-                        }
+                        } while (++i < entries.size());
                         newEntries.add(e);
-                        if (compareAndSet(entries, newEntries)) {
-                            if (replacedEntry != null) {
-                                replacedEntry.cancelExpiration();
-                            }
+                        if (compareAndSet(entries, Collections.unmodifiableList(newEntries))) {
+                            scheduleCacheExpirationIfNeeded(ttl, loop);
                             return;
                         }
                     } else if (compareAndSet(entries, Collections.singletonList(e))) {
+                        scheduleCacheExpirationIfNeeded(ttl, loop);
                         return;
                     }
                 }
             } else {
-                List<DefaultDnsCacheEntry> entries = getAndSet(Collections.singletonList(e));
-                cancelExpiration(entries);
+                set(Collections.singletonList(e));
+                scheduleCacheExpirationIfNeeded(ttl, loop);
+            }
+        }
+
+        private void scheduleCacheExpirationIfNeeded(int ttl, EventLoop loop) {
+            long start = System.nanoTime();
+            long ttlNanos = TimeUnit.SECONDS.toNanos(ttl);
+
+            for (;;) {
+                ScheduledFuture<?> oldFuture = FUTURE_UPDATER.get(this);
+                long calculatedTTL = ttlNanos - (System.nanoTime() - start);
```

Comment:
extension of 
https://github.com/netty/netty/pull/8187#discussion_r209771462 ... I'm not convinced this is really necessary as the TTL granularity is in seconds and isn't tightly synchronized anyways. The cost to call `nanoTime` is not "free" either and this adds complexity. My vote would be to remove it and simplify for now unless this is proven to be a problem.

If we do think this is necessary we should ensure we only call `nanoTime` once per iteration by introducing a local variable to store the "last result in". This code currently calls `nanoTime` twice for the first loop iteration.

/cc @normanmaurer @carl-mastrangelo

Reply:
I think I am with @Scottmitch here.. Let me revert the change and add a comment.